Scene unique nodes: clarify intended usage (#6670)
godotengine/godot-docs: tutorials/scripting/scene_unique_nodes.rst

Introduction
------------

> There are times in a project where a node needs to be referred to by

Author: Overall I think the intro is kind of unusual, implying that node to node references are something noteworthy/rare. But intros are hard. 😄 Agree with both suggestions as improvements.

> a script. However, the node's position in the tree might change

Author: Taking suggestion for now, and I'll look at redoing this sentence. Allowing "you" in the doc will I think make it clearer--these moves don't happen out of nowhere.

.. image:: img/percent.png

To use a unique node in a script, use the ``%`` symbol and the node's

> name to create a path for ``get_node()``. For example:

Reviewer: This sentence is badly phrased, and I don't think your suggestion helps, quite the opposite. How about:

> You can now use the node in your script. For example, you can reference it with
a ``get_node()`` method by typing the ``%`` symbol, followed by the node's name:

Author: I think it's clearer as far as technical correctness goes, but a rework is better. I like your suggestion, will apply it.

 .. code-tab:: csharp

    GetNode<Button>("%RedButton").Text = "Hello";

Same-scene limitation
---------------------

A scene unique node can only be retrieved by a node inside the same scene.

If the same scene is instantiated twice, a node inside the first instance
calling ``get_node("%RedButton")`` will get the button in the first instance,
not the button in the second instance.

This limitation is perfect for a UI scene like a health bar, where each instance
of the health bar must not interact with other instances.

When a scene is instantiated inside another scene, the two scenes can't access

> scene unique nodes from each other. If a Player scene contains an instance of a

Author: Note from a discussion in Discord where scene unique nodes came up again: you can access the scene unique nodes, but you have to navigate into the scene first. Like `get_node("Hand/Sword/%Hilt")`, or `s = get_node("Hand/Sword")` `s.get_node("%Hilt")`. Seems worth noting, to me.

 .. code-tab:: csharp

    GetNode<Button>("%RedButton").Text = "Hello";

Same-scene limitation
---------------------

> A scene unique node can only be retrieved by a node inside the same scene.

Reviewer: ```suggestion
A scene unique node can only be retrieved by a node inside of the same scene.
```

Author: That doesn't sound right to me (AmE). I don't know enough grammar terminology to say exactly why, though. 😄 A couple online grammar checkers say the `of` is not necessary if I paste in this sentence.

If the same scene is instantiated twice, a node inside the first instance
calling ``get_node("%RedButton")`` will get the button in the first instance,
not the button in the second instance.

This limitation is perfect for a UI scene like a health bar, where each instance

> of the health bar must not interact with other instances.

Reviewer: I'm not sure this is worth a mention. This is just how node paths work. You can right the same for any node path within two instances of the same scene. Maybe there is a point to make here, but I'm only confused by the focus of these two paragraphs.

Author: The point is that if the scene unique name were global, then the two instances would interact. The implications of having scene unique nodes cached in each scene aren't obvious to everybody, so it seems worth spelling this out a little. And after understanding all this, someone still might want to file a proposal to make scene unique nodes global, so I'm trying to predict that a bit and show why this behavior is actually desirable.

Author: Ended up removing. It actually might be nifty to support getting a scene unique node from a nested scene, who knows?

Also removed the next paragraph. Doesn't seem necessary to word it all out with the example below.

When a scene is instantiated inside another scene, the nodes in the outer scene
can't access scene unique nodes in the inner scene using the scene unique name.
Likewise, nodes in the inner scene can't access scene unique nodes in the outer
scene using the scene unique name.

> To demonstrate this limitation, consider this example Player scene that
> instances a Sword scene:

Reviewer: ```suggestion
To demonstrate this limitation, consider this example **Player** scene that
instances a **Sword** scene:
```

.. image:: img/unique_name_scene_instance_example.png

> Here are the results of ``get_node()`` calls inside the Player script:
> 
> - ``get_node("%Eyes")`` returns the Eyes node.
> - ``get_node("%Hilt")`` returns ``null``.
> 
> These are the results of ``get_node()`` calls inside the Blade script:
> 
> - ``get_node("%Eyes")`` returns ``null``.
> - ``get_node("%Hilt")`` returns the Hilt node.

Reviewer: ```suggestion
Here are the results of ``get_node()`` calls inside the **Player** script:

- ``get_node("%Eyes")`` returns the **Eyes** node.
- ``get_node("%Hilt")`` returns ``null``.

These are the results of ``get_node()`` calls inside the **Blade** script:

- ``get_node("%Eyes")`` returns ``null``.
- ``get_node("%Hilt")`` returns the **Hilt** node.
```

If a script has access to a node in another scene, it can call ``get_node()`` on
that node to get scene unique nodes from that node's scene. This also works in a
node path, which avoids multiple ``get_node()`` calls. Here are two ways to get

> the Hilt node from the Player script using scene unique nodes:
> 
> - ``get_node("Hand/Sword").get_node("%Hilt")`` returns the Hilt node.
> - ``get_node("Hand/Sword/%Hilt")`` also returns the Hilt node.

Reviewer: ```suggestion
the **Hilt** node from the **Player** script using scene unique nodes:

- ``get_node("Hand/Sword").get_node("%Hilt")`` returns the **Hilt** node.
- ``get_node("Hand/Sword/%Hilt")`` also returns the **Hilt** node.
```

Scene unique names don't only work at the end of a node path. They can be used
in the middle to navigate from one node to another. For example, the Sword node

> is marked as a scene unique node in the Player scene, so this is possible:

Reviewer: ```suggestion
is marked as a scene unique node in the **Player** scene, so this is possible:
```

> - ``get_node("%Sword/%Hilt")`` returns the Hilt node.

Reviewer: ```suggestion
- ``get_node("%Sword/%Hilt")`` returns the **Hilt** node.
```

Alternatives
------------

Scene unique nodes are a useful tool to navigate a scene. However, there are
some situations where other techniques may be better.

A :ref:`Group <doc_groups>` allows locating a node from any other node, no
matter what scene the two nodes are located in. Scene unique nodes are usually

> not suitable to find a Player node from an Enemy node, for example, but a group
> with the Player node in it is a typical solution.

Reviewer: ```suggestion
not suitable to find a **Player** node from an **Enemy** node, for example, but a group
with the **Player** node in it is a typical solution.
```

I'm not sure that a group for just one Player node is actually a typical solution. It would make sense for all enemies to be in a group, but then the example doesn't make much sense for scene unique nodes. Whatever, I guess.

Author: I took this example from the PR discussion: https://github.com/godotengine/godot/pull/60298#issuecomment-1100637400. Better example would definitely be nice, but I haven't thought of anything that would be universally understood without some setup. Maybe it's not worth trying to have an example here, and just rely on people clicking through to the groups page without this extra convincing. 😄

Author: Removed the example.

A :ref:`Singleton (AutoLoad) <doc_singletons_autoload>` is an always loaded node

> that can easily be accessed by any node regardless of scene. These are useful

Reviewer: ```suggestion
that can be accessed directly by any node regardless of the scene. These are useful
```

when some data or functionality is shared globally.

:ref:`Node.find_child() <class_Node_method_find_child>` finds a node by name
without knowing its full path. This seems similar to a scene unique node, but
this method is able to find nodes in nested scenes, and doesn't require marking
the node in the scene editor in any way. However, this method is slow. Scene

> unique nodes are cached by Godot and are fast to retrieve, but ``find_child()``
> needs to search the tree while the game is running.

Reviewer: ```suggestion
unique nodes are cached by Godot and are fast to retrieve, but ``find_child()``
needs to check every child and grandchild each time the method is called.
```

Author: This is tidy, but I don't like leaving it at grandchild without any hint that it goes forever. Maybe:

```suggestion
unique nodes are cached by Godot and are fast to retrieve, but each time the method is called, ``find_child()``
needs to check every descendant (every child, grandchild, and so on).
```

Same-scene limitation
---------------------

A scene unique node can only be retrieved by a node inside the same scene. To
demonstrate this limitation, consider this example **Player** scene that
instances a **Sword** scene:

.. image:: img/unique_name_scene_instance_example.png

Here are the results of ``get_node()`` calls inside the **Player** script:

- ``get_node("%Eyes")`` returns the **Eyes** node.
- ``get_node("%Hilt")`` returns ``null``.

These are the results of ``get_node()`` calls inside the **Blade** script:

- ``get_node("%Eyes")`` returns ``null``.
- ``get_node("%Hilt")`` returns the **Hilt** node.

If a script has access to a node in another scene, it can call ``get_node()`` on
that node to get scene unique nodes from that node's scene. This also works in a
node path, which avoids multiple ``get_node()`` calls. Here are two ways to get
the **Hilt** node from the **Player** script using scene unique nodes:

- ``get_node("Hand/Sword").get_node("%Hilt")`` returns the **Hilt** node.
- ``get_node("Hand/Sword/%Hilt")`` also returns the **Hilt** node.

Scene unique names don't only work at the end of a node path. They can be used

> in the middle to navigate from one node to another. For example, the Sword node

Author: ```suggestion
in the middle to navigate from one node to another. For example, the **Sword** node
```
